core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/middleware.rs

```diff
@@ -0,0 +1,394 @@
+    transport: Transport,
+    _guard: GaugeGuard,
+}
+
+impl<S> LimitMiddleware<S> {
+    pub(crate) fn new(inner: S, requests_per_minute_limit: Option<NonZeroU32>) -> Self {
+        Self {
+            inner,
+            rate_limiter: requests_per_minute_limit
+                .map(|limit| RateLimiter::direct(Quota::per_minute(limit))),
+            transport: Transport::Ws,
+            _guard: API_METRICS.ws_open_sessions.inc_guard(1),
+        }
+    }
+}
+
+impl<'a, S> RpcServiceT<'a> for LimitMiddleware<S>
+where
+    S: Send + Sync + RpcServiceT<'a>,
+{
+    type Future = ResponseFuture<S::Future>;
+
+    fn call(&self, request: Request<'a>) -> Self::Future {
+        if let Some(rate_limiter) = &self.rate_limiter {
+            let num_requests = NonZeroU32::MIN; // 1 request, no batches possible
+
+            // Note: if required, we can extract data on rate limiting from the error.
+            if rate_limiter.check_n(num_requests).is_err() {
+                METRICS.rate_limited[&self.transport].inc();
+
+                let rp = MethodResponse::error(
+                    request.id,
+                    ErrorObject::borrowed(
+                        ErrorCode::ServerError(
+                            reqwest::StatusCode::TOO_MANY_REQUESTS.as_u16().into(),
+                        )
+                        .code(),
+                        "Too many requests",
+                        None,
+                    ),
+                );
+                return ResponseFuture::ready(rp);
+            }
+        }
+        ResponseFuture::future(self.inner.call(request))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MetadataMiddleware<S> {
+    inner: S,
+    registered_method_names: Arc<HashSet<&'static str>>,
+    #[cfg(test)]
+    call_tracer: CallTracer,
+}
+
+impl<S> MetadataMiddleware<S> {
+    pub fn new(inner: S, registered_method_names: Arc<HashSet<&'static str>>) -> Self {
+        Self {
+            inner,
+            registered_method_names,
+            #[cfg(test)]
+            call_tracer: CallTracer::default(),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn with_call_tracer(mut self, call_tracer: Option<CallTracer>) -> Self {
+        if let Some(call_tracer) = call_tracer {
+            self.call_tracer = call_tracer;
+        }
+        self
+    }
+}
+
+impl<'a, S> RpcServiceT<'a> for MetadataMiddleware<S>
+where
+    S: Send + Sync + RpcServiceT<'a>,
+{
+    type Future = WithMethodMetadata<S::Future>;
+
+    fn call(&self, request: Request<'a>) -> Self::Future {
+        // "Normalize" the method name by searching it in the set of all registered methods. This extends the lifetime
+        // of the name to `'static` and maps unknown methods to "", so that method name metric labels don't have unlimited cardinality.
+        let method_name = self
+            .registered_method_names
+            .get(request.method_name())
```

Comment:
@RomanBrodetski Are we OK with this? I think it would be theoretically possible to map method names to the old format, but I wouldn't want to do that.

Reply:
ohh this is a little unfortunate: I have no problem changing the dashboards, but we won't have an easy way to see historic trends. However I think the overall change is good so let's proceed